### PR TITLE
Execute e2e tests inside their own pid and mount namespaces.

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

- Execute e2e tests inside their own pid and mount namespaces:
  - pid namespace ensure no pending process will live after tests finished
  - mount namespace will allow to manipulate binds without to worry about host mount namespace, and it allow actually to setup temporary home directories
- A temporary filesystem is now setup and home directories are
created in this temporary filesystem and bound to the original
ones, that would ease e2e tests and avoid to screw them up during
tests execution (think about singularity cache, library tokens, PGP
keys ...)
- Remove unnecessary call to `EnsureImage`, cleanup code a bit.

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: e2e/internal/e2e/docker.go [hunk 1]
```diff
@@ -50,6 +51,13 @@ func PrepRegistry(t *testing.T, env TestEnv) {
 			ExpectExit(0),
 		)
 
+		// sleep a bit to let time to registry to start
+		// and setup everything. The start script already
+		// sleep for 2 seconds before pushing a busybox image in
```
Comment (WestleyK) [inline e2e/internal/e2e/docker.go]: It looks like its sleeping for 5 seconds

```suggestion
		// sleep for 5 seconds before pushing a busybox image in
```

**Review by WestleyK — COMMENTED** [review]
Looks good to me, just one suggestion...

File: e2e/internal/e2e/docker.go (hunk 1 above)
Comment (cclerget) [reply e2e/internal/e2e/docker.go]: Oh I refer to this line https://github.com/sylabs/singularity/blob/master/e2e/testdata/Docker_registry.def#L19

Comment (cclerget) [reply e2e/internal/e2e/docker.go]: I will mention the file to avoid confusion

**Comment by jscook2345** [comment]
Can I ask why you want to do this? CircleCI already runs tests in a VM or a container. Is this for the local case where you're running `make test` on your local linux system?

**Comment by cclerget** [comment]
> Can I ask why you want to do this? CircleCI already runs tests in a VM or a container. Is this for the local case where you're running `make test` on your local linux system?

Yes essentially, additionally user/root homes are replaced by an empty one where e2e tests could do what they needs : play with cache, PGP keys, token, running/stopping all instances ..., while this is fine for CI where we start from scratch, this is less good on a dev machine where some image are already cached that I would like to keep, PGP keys that I need to keep, prevent few instances currently running from being stopped by a test executing `instance stop --all`. Also I would like to not mess with host installation and/or other tests by using a custom temporary binding on top `/usr/local/etc/singularity` to test different directives value in `singularity.conf`.

File: e2e/internal/e2e/docker.go [hunk 2]
```diff
@@ -50,6 +51,13 @@ func PrepRegistry(t *testing.T, env TestEnv) {
 			ExpectExit(0),
 		)
 
+		// Sleep a bit to let time to registry to start and setup
+		// everything. The start script already sleep for 2 seconds
+		// (in e2e/testdata/Docker_registry.def) before pushing a
+		// busybox image in the docker registry. So it's safer to
+		// sleep few seconds at this point
+		time.Sleep(5 * time.Second)
```
Comment (mem) [inline e2e/internal/e2e/docker.go]: Sleeping is inherently brittle. Is there something we could try (multiple times, sleeping a little in between if needed) to test that the registry is up or not? From quickly browsing the docs, I see that contacting the login end point should be sufficient, as it should reply with either 200 or 401 if the registry is up (in other words, you don't need to login, you just need to try it).

File: e2e/internal/e2e/home_linux.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+// SetupHomeDirectories creates temporary home directories for
+// privileged and unprivileged users and bind mount those directories
+// on top of real ones. It's possible because e2e tests are executed
+// in a dedicated mount namespace.
+func SetupHomeDirectories(t *testing.T) {
+	var unprivUser, privUser *user.User
+
+	sessionDir := buildcfg.SESSIONDIR
+	unprivUser = CurrentUser(t)
```
Comment (mem) [inline e2e/internal/e2e/home_linux.go:36]: seeing how the code below depends on this really being a different user, shouldn't this code assert that 1) this is in fact an unprivileged user; 2) that this is different from the privileged user below?

File: e2e/internal/e2e/home_linux.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		if err := os.Chown(unprivHome, int(unprivUser.UID), int(unprivUser.GID)); err != nil {
+			t.Fatalf("failed to set temporary home %s owner: %s", unprivHome, err)
+		}
+		if err := os.Mkdir(privHome, 0700); err != nil {
+			t.Fatalf("failed to create temporary home %s: %s", privHome, err)
+		}
+
+		// re-create current working directory if tests are run from the user
+		// home directory and bind it
+		if strings.HasPrefix(cwd, unprivRealHome) {
```
Comment (mem) [inline e2e/internal/e2e/home_linux.go]: If you where to give this bit of logic a name, it would be easier to read.

Let's see if I can unravel this correctly.

Let's say

```
sessionDir := "/a/b"
unprivUser.Name = "joe"
unprivHome := "/a/b/joe"
unprivRealHome := "/x/y"
cwd := "/x/y/z/1/2/3"    // filepath.Dir(cwd) == "/x/y/z/1/2"
trimmed := "/z/1/2"
path := "/a/b/joe/z/1/2"
```

correct?

so it's extracting (part of) the current working directory relative to the home directory of the current (?) user and reparenting on top of the temporary home?

And why is this necessary only if the current working directory is within the current user's home? (why is there a condition on line 83)

Also, what's special about the directory "3" above? Why it is kept out of this whole thing?

File: e2e/internal/e2e/home_linux.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			t.Fatalf("failed to bind %s to %s: %s", unprivHome, unprivRealHome, err)
+		}
+		if err := syscall.Mount(privHome, privRealHome, "", syscall.MS_BIND, ""); err != nil {
+			t.Fatalf("failed to bind %s to %s: %s", privHome, privRealHome, err)
+		}
+		if err := os.Chdir(cwd); err != nil {
+			t.Fatalf("failed to change directory to %s: %s", cwd, err)
+		}
+
+		// create .rpmmacros files
```
Comment (mem) [inline e2e/internal/e2e/home_linux.go]: up to this point everything was pretty generic.

What is this bit about? Seems pretty unrelated to setting up home directories.

File: e2e/internal/e2e/home_linux.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		// finally bind temporary homes on top of real ones
+		// in order to not screw them by accident during e2e
+		// tests execution
+		if err := syscall.Mount(unprivHome, unprivRealHome, "", syscall.MS_BIND|syscall.MS_REC, ""); err != nil {
+			t.Fatalf("failed to bind %s to %s: %s", unprivHome, unprivRealHome, err)
+		}
+		if err := syscall.Mount(privHome, privRealHome, "", syscall.MS_BIND, ""); err != nil {
+			t.Fatalf("failed to bind %s to %s: %s", privHome, privRealHome, err)
+		}
+		if err := os.Chdir(cwd); err != nil {
```
Comment (mem) [inline e2e/internal/e2e/home_linux.go:108]: oh, this is subtle. Please add a comment as to why this is needed.

File: e2e/internal/e2e/home_linux.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+// privileged and unprivileged users and bind mount those directories
+// on top of real ones. It's possible because e2e tests are executed
+// in a dedicated mount namespace.
+func SetupHomeDirectories(t *testing.T) {
+	var unprivUser, privUser *user.User
+
+	sessionDir := buildcfg.SESSIONDIR
+	unprivUser = CurrentUser(t)
+
+	Privileged(func(t *testing.T) {
```
Comment (mem) [inline e2e/internal/e2e/home_linux.go:38]: just a general comment about this anonymous function: it's doing almost the same thing twice, once for the privileged case and once for the unprivileged case. Since there's no real logic overlap between the two cases, the code would be easier to read if you did one part first (say the privileged one) and the other part second (say the unprivileged one).

File: e2e/internal/e2e/init/init_linux.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+// supplied PID.
+static pid_t getProcInfo(pid_t pid, uid_t *uid, gid_t *gid) {
+	FILE *status;
+	char procPath[SIZE];
+	char *line = NULL;
+	size_t len = 0;
+	pid_t ppid = 1;
+
+	memset(procPath, 0, SIZE);
+	snprintf(procPath, SIZE-1, "/proc/%d/status", pid);
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go]: this is probably never going to be the case with this code, but just to be correct, please check the return value of snprintf to make sure that the string did in fact fit in the provided buffer.

File: e2e/internal/e2e/init/init_linux.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	status = fopen(procPath, "r");
+	if ( status == NULL ) {
+		return 1;
+	}
+
+	while ( getline(&line, &len, status) != -1 ) {
+		if ( ppid == 1 ) {
+			sscanf(line, "PPid:\t%d", &ppid);
+		}
+		if ( *uid == 0 ) {
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go:47]: since this code assumes that uid == 0 before entering this function, shouldn't it check that that's the case?

File: e2e/internal/e2e/init/init_linux.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+			sscanf(line, "Uid:\t%d", uid);
+		}
+		if ( *gid == 0 ) {
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go:50]: same here

File: e2e/internal/e2e/init/init_linux.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+	free(line);
+	fclose(status);
+
+	return ppid;
+}
+
+// getUnprivIDs searches recursively up the process parent chain to find a
+// process with a non-root UID, then returns the UID and GID of that process.
+static int getUnprivIDs(pid_t pid, uid_t *uid, gid_t *gid) {
+	if ( pid == 1 ) {
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go:67]: This is quite subtle as a terminating condition. At first I thought it was just a special case but then I realized that things are wired up so that getUnprivIDs returns 1 on errors. So this bit of code is playing two roles: if the code is run by root, it will terminate here because eventually it will walk up to PID 1; if the code is not run by root, but there's an error, it will terminate here because of the error.

Can you please add a comment?

File: e2e/internal/e2e/init/init_linux.go [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+		return -1;
+	}
+	pid_t ppid = getProcInfo(pid, uid, gid);
+	if ( *uid == 0 ) {
+		return getUnprivIDs(ppid, uid, gid);
+	}
+	return 0;
+}
+
+__attribute__((constructor)) static void init(void) {
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go]: can you please add a comment as to why this needs to be set up line like this? (I'm talking about the attribute constructor). This construction makes reading the code really hard, because you start at e2e/suite.go, where `func main` is, and you can continue reading without ever noticing that this thing ran first.

I would say _this_ function needs a comment and `func main` also needs a comment refering this.

File: e2e/internal/e2e/init/init_linux.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	// use a mount namespace in order to bind a temporary filesystem
+	// on top of home directories and not screw them up by accident
+	// during tests execution
+	if ( unshare(CLONE_FS) < 0 ) {
+		fprintf(stderr, "failed to unshare filesystem: %s\n", strerror(errno));
+		exit(1);
+	}
+	if ( unshare(CLONE_NEWNS) < 0 ) {
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go]: since this requires root (or CAP_SYS_ADMIN, same difference), maybe check for that before trying to do this in order to produce a better error message. The error message below is pretty generic and it will make things hard to debug for other people.

File: e2e/internal/e2e/init/init_linux.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		fprintf(stderr, "failed to set private mount propagation: %s\n", strerror(errno));
+		exit(1);
+	}
+	// use a PID namespace if possible to avoid to left some processes
+	// once tests are done. Child process won't catch orphaned child processes like
+	// instances, and we can't really catch them correctly to avoid conflicts during
+	// `cmd.Wait()` calls. But this is not a big deal compared to detached processes
+	// that could keep running on host machine after the tests execution.
+	if ( unshare(CLONE_NEWPID) == 0 ) {
+		pid_t forked = fork();
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go:84]: the body of this if is independent of the rest of the code. Please move it to its own function, just for readability's sake.

I would even say everything from the unshare call down should be its own function.

File: e2e/internal/e2e/init/init_linux.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+		if ( forked > 0 ) {
+			// parent process will wait that tests execution finished
+			int status, exit_status = 0;
+			pid_t child;
+
+			child = waitpid(forked, &status, 0);
+			if ( child < 0 ) {
+				fprintf(stderr, "unexpected error while waiting childs: %s\n", strerror(errno));
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go]:
```suggestion
				fprintf(stderr, "unexpected error while waiting for children: %s\n", strerror(errno));
```

File: e2e/internal/e2e/init/init_linux.go [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	if ( getUnprivIDs(getppid(), &uid, &gid) < 0 ) {
+		fprintf(stderr, "failed to retrieve user information\n");
+		exit(1);
+	}
+	if ( uid == 0 || gid == 0 ) {
+		fprintf(stderr, "failed to retrieve user information\n");
+		exit(1);
+	}
+
+	// use a mount namespace in order to bind a temporary filesystem
```
Comment (mem) [inline e2e/internal/e2e/init/init_linux.go]: the bit between here and line 97 could be moved to its own function (setup_mount_namespace or some such).

**Review by mem — CHANGES_REQUESTED** [review]

File: e2e/internal/e2e/docker.go (hunk 2 above)
Comment (cclerget) [reply e2e/internal/e2e/docker.go]: Yeah, I will check how to do that better

File: e2e/internal/e2e/home_linux.go (hunk 3 above)
Comment (cclerget) [reply e2e/internal/e2e/home_linux.go:36]: Can avoid that by simply adding a root check in constructor in `init_linux.go`

File: e2e/internal/e2e/home_linux.go (hunk 4 above)
Comment (cclerget) [reply e2e/internal/e2e/home_linux.go]: > correct?
so it's extracting (part of) the current working directory relative to the home directory of the current (?) user and reparenting on top of the temporary home?

Right

> And why is this necessary only if the current working directory is within the current user's home? (why is there a condition on line 83)

As we use go module, the source could be everywhere, if the source base is in `/tmp/singularity` we don't need to bind CWD on top of the user home directory, but if source code is indeed somewhere in user directory we need to bind it on top of the temporary one in order to execute tests correctly.

> Also, what's special about the directory "3" above? Why it is kept out of this whole thing?

It would have been easier to have something like `buildcfg.SOURCEDIR` in order to determine the source code location, but as tests are executed relative to the directory where `_test.go` is located, so in `$SOURCEDIR/e2e`, `filepath.Dir(cwd)` is equivalent to source directory

File: e2e/internal/e2e/home_linux.go (hunk 5 above)
Comment (cclerget) [reply e2e/internal/e2e/home_linux.go]: `SetupHomeDirectories` and everything they can/need to contain

File: e2e/internal/e2e/home_linux.go (hunk 6 above)
Comment (cclerget) [reply e2e/internal/e2e/home_linux.go:108]: will do

File: e2e/internal/e2e/home_linux.go (hunk 7 above)
Comment (cclerget) [reply e2e/internal/e2e/home_linux.go:38]: I think that would make the code more complicated, for the unprivileged part there is also privileged operation to execute and that could be confusing

File: e2e/internal/e2e/init/init_linux.go (hunk 8 above)
Comment (cclerget) [reply e2e/internal/e2e/init/init_linux.go]: Will do

File: e2e/internal/e2e/init/init_linux.go (hunk 9 above)
Comment (cclerget) [reply e2e/internal/e2e/init/init_linux.go:47]: Not required, `uid` and `gid` are initialized to 0 : https://github.com/sylabs/singularity/blob/11503481784c21a6637e59afe239f7494e71d2ee/e2e/internal/e2e/init/init_linux.go#L71

and there is already a check at https://github.com/sylabs/singularity/blob/11503481784c21a6637e59afe239f7494e71d2ee/e2e/internal/e2e/init/init_linux.go#L64

File: e2e/internal/e2e/init/init_linux.go (hunk 11 above)
Comment (cclerget) [reply e2e/internal/e2e/init/init_linux.go:67]: will do

File: e2e/internal/e2e/init/init_linux.go (hunk 13 above)
Comment (cclerget) [reply e2e/internal/e2e/init/init_linux.go]: will add a generic check at the beginning for root user

File: e2e/internal/e2e/init/init_linux.go (hunk 12 above)
Comment (cclerget) [reply e2e/internal/e2e/init/init_linux.go]: will do

File: e2e/e2e_test.go [hunk 17]
```diff
@@ -7,6 +7,16 @@ package e2e
 
 import (
 	"testing"
+
+	// This import will execute a CGO section with the help of a C
+	// constructor section "init". As we always requires to run e2e
```
Comment (gvallee) [inline e2e/e2e_test.go]:
```suggestion
	// constructor section "init". As we always require to run e2e
```

File: e2e/e2e_test.go [hunk 18]
```diff
@@ -7,6 +7,16 @@ package e2e
...
+	// tests as root, the C part is responsible of finding original
```
Comment (gvallee) [inline e2e/e2e_test.go]:
```suggestion
	// tests as root, the C part is responsible of finding the original
```

File: e2e/e2e_test.go [hunk 19]
```diff
@@ -7,6 +7,16 @@ package e2e
...
+	// user which execute tests, it will also create a dedicated pid
```
Comment (gvallee) [inline e2e/e2e_test.go]:
```suggestion
	// user who executes tests; it will also create a dedicated pid
```

File: e2e/e2e_test.go [hunk 20]
```diff
@@ -7,6 +7,16 @@ package e2e
...
+	// and mount namespace for e2e tests, and will finally restore
+	// identity to the original user but will retain privileges for
+	// Privileged method allowing to execute a function with root
```
Comment (gvallee) [inline e2e/e2e_test.go]:
```suggestion
	// Privileged methods enabling the execution of a function with root
```

File: e2e/internal/e2e/docker.go [hunk 21]
```diff
@@ -50,6 +52,25 @@ func PrepRegistry(t *testing.T, env TestEnv) {
 			ExpectExit(0),
 		)
 
+		// start script in e2e/testdata/Docker_registry.def will listen
+		// on port 5111 once docker registry is up and initialized, so
+		// we are trying to connect to this port until we got a response,
+		// without any response after 10 seconds we abort tests execution
+		// because start script probably failed
```
Comment (gvallee) [inline e2e/internal/e2e/docker.go]:
```suggestion
		// because the start script probably failed
```

File: e2e/internal/e2e/init/init_linux.go [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+		return -1;
+	}
+	pid_t ppid = getProcInfo(pid, uid, gid);
+	if ( *uid == 0 ) {
+		return getUnprivIDs(ppid, uid, gid);
+	}
+	return 0;
+}
+
+// create and use a PID namespace if possible to avoid to left some processes
```
Comment (gvallee) [inline e2e/internal/e2e/init/init_linux.go]:
```suggestion
// create and use a PID namespace if possible to avoid leaving some processes
```

**Review by gvallee — CHANGES_REQUESTED** [review]
LGTM, I only change a few things in a few comments.

**Review by gvallee — APPROVED** [review]
LGTM!

**Review by WestleyK — APPROVED** [review]
Looks good to me :+1:

File: e2e/internal/e2e/home_linux.go (hunk 4 above)
Comment (mem) [reply e2e/internal/e2e/home_linux.go]: > As we use go module, the source could be everywhere, if the source base is in `/tmp/singularity` we don't need to bind CWD on top of the user home directory, but if source code is indeed somewhere in user directory we need to bind it on top of the temporary one in order to execute tests correctly.

Ah, got it. That makes sense.

> > Also, what's special about the directory "3" above? Why it is kept out of this whole thing?
> 
> It would have been easier to have something like `buildcfg.SOURCEDIR` in order to determine the source code location, but as tests are executed relative to the directory where `_test.go` is located, so in `$SOURCEDIR/e2e`, `filepath.Dir(cwd)` is equivalent to source directory

I think I understand what you are saying.

I would not mind adding SOURCEDIR to buildcfg (I actually would need this to test the plugin stuff; the closest right now is buildcfg.BUILDDIR, which by default contains a directory one deeper than SOURCEDIR, but that can be changed). You could think of it as a debugging tool, it's pretty common to record that kind of information.

If the reason for this logic is getting to SOURCEDIR from e2e, I would prefer to make that change (add SOURCEDIR to buildcfg).

**Review by mem — COMMENTED** [review]
This looks good to me. If you could add buildcfg.SOURCEDIR, that would be great.

**Comment by cclerget** [comment]
> This looks good to me. If you could add buildcfg.SOURCEDIR, that would be great.

@mem Yep, added

**Review by mem — APPROVED** [review]
OK, let's go with this.
